File: org.dcs.nifi.processors/src/main/java/org/dcs/nifi/processors/ProcessorUtils.java
package org.dcs.nifi.processors;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.Set;

import org.apache.nifi.components.PropertyDescriptor;
import org.apache.nifi.components.PropertyDescriptor.Builder;
import org.apache.nifi.components.Validator;
import org.apache.nifi.processor.Relationship;
import org.dcs.api.service.FlowModuleConstants;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class ProcessorUtils {

	private static final Logger logger = LoggerFactory.getLogger(ProcessorUtils.class);

	public static List<PropertyDescriptor> generatePropertyDescriptors(Map<String, Properties> propertiesMap) {
		List<PropertyDescriptor> propertyDescriptors = new ArrayList<>();
		for(Properties props : propertiesMap.values()) {
			Builder propertyBuilder = new PropertyDescriptor.Builder();
			for(Map.Entry<Object, Object> entrySet : props.entrySet()) {
				addPropertyToDescriptor(propertyBuilder, entrySet.getKey().toString(), entrySet.getValue().toString());
			}
			propertyBuilder.addValidator(Validator.VALID);
			propertyDescriptors.add(propertyBuilder.build());
		}
		return propertyDescriptors;
	}

	public static void addPropertyToDescriptor(Builder propertyBuilder, String key, String value) {
		switch(key) {
		case FlowModuleConstants.PROPERTY_NAME:
			propertyBuilder.name(value);
			break;
		case FlowModuleConstants.PROPERTY_DESCRIPTION:
			propertyBuilder.description(value);
			break;
		case FlowModuleConstants.PROPERTY_REQUIRED:
			propertyBuilder.required(Boolean.valueOf(value));
			break;
		case FlowModuleConstants.PROPERTY_DEFAULT_VALUE:
			propertyBuilder.defaultValue(value);
			break;
		default:
			logger.warn("Property " + key + " not known");
		}	
	}

	public static Set<Relationship> generateRelationships(Map<String, Properties> relationshipMap) {
		Set<Relationship> relationships = new HashSet<>();
		for(Properties props : relationshipMap.values()) {
			org.apache.nifi.processor.Relationship.Builder relationshipBuilder = new Relationship.Builder();
			for(Map.Entry<Object, Object> entrySet : props.entrySet()) {
				addRelationship(relationshipBuilder, entrySet.getKey().toString(), entrySet.getValue().toString());
			}			
			relationships.add(relationshipBuilder.build());
		}
		return relationships;
	}
	
	public static void addRelationship(org.apache.nifi.processor.Relationship.Builder relationshipBuilder, 
			String key, 
			String value) {
		switch(key) {
		case FlowModuleConstants.RELATIONSHIP_NAME:
			relationshipBuilder.name(value);
			break;
		default:
			logger.warn("Relationship " + key + " not known");
		}	
	}
	
	public static Properties generateValueProperties(Map<PropertyDescriptor, String> propertyDescriptorValueMap) {
		Properties propertiesValueMap = new Properties();
		for(Map.Entry<PropertyDescriptor, String> propertyDescriptorValue : propertyDescriptorValueMap.entrySet()) {
			propertiesValueMap.setProperty(propertyDescriptorValue.getKey().getName(), propertyDescriptorValue.getValue());
		}
		return propertiesValueMap;
	}
	
	public static Relationship getSuccessRelationship(Set<Relationship> relationships) {
		for(Relationship relationship : relationships) {
			if(FlowModuleConstants.REL_SUCCESS_ID.equals(relationship.getName())) {
				return relationship;
			}
		}
		return null;
	}
}

File: org.dcs.nifi.processors/src/main/scala/org/dcs/nifi/processors/ProcessorUtils.scala
package org.dcs.nifi.processors

import org.apache.nifi.processor.Relationship
import org.dcs.api.service.FlowModule
import org.apache.nifi.components.PropertyDescriptor
import org.apache.nifi.processor.Relationship.{ Builder => RelationshipBuilder }
import org.apache.nifi.components.PropertyDescriptor.{ Builder => PropertyDescriptorBuilder }
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.Properties
import java.util.{Map => JavaMap}
import java.util.{Set => JavaSet}
import scala.collection.mutable.{ Map => MutableMap }
import scala.collection.JavaConverters._
import org.apache.nifi.components.Validator



object ProcessorUtils {
  val logger: Logger = LoggerFactory.getLogger(classOf[ProcessorUtils])
  
  def generatePropertyDescriptors(propDescMap: JavaMap[String, JavaMap[String, String]]): List[PropertyDescriptor] = {
    var propertyDescriptors: List[PropertyDescriptor] = Nil
    for ((k, v) <- propDescMap.asScala) {
      val propDescBuilder: PropertyDescriptorBuilder = new PropertyDescriptorBuilder()
      v.asScala.foreach { x =>
        addPropertyDescriptor(propDescBuilder, x._1, x._2)
      }
      propDescBuilder.addValidator(Validator.VALID)
      propertyDescriptors = propDescBuilder.build() :: propertyDescriptors
    }
    propertyDescriptors
  }

  def addPropertyDescriptor(propDescBuilder: PropertyDescriptorBuilder,
                            key: String,
                            value: String) = key match {
    case FlowModule.PropertyName => propDescBuilder.name(value)
    case FlowModule.PropertyDescription=> propDescBuilder.description(value)
    case FlowModule.PropertyRequired => propDescBuilder.required(value.toBoolean)
    case FlowModule.PropertyDefaultValue => propDescBuilder.defaultValue(value)
    case _ => logger.warn("Relationship " + key + " not known")
  }

  def generateRelationships(relationshipMap: JavaMap[String, JavaMap[String, String]]): Set[Relationship] = {
    var relationships: Set[Relationship] = Set()
    for ((k, v) <- relationshipMap.asScala) {
      val relationshipBuilder: RelationshipBuilder = new RelationshipBuilder()
      v.asScala.foreach { x =>
        addRelationship(relationshipBuilder, x._1, x._2)
      }
      relationships += relationshipBuilder.build()
    }
    relationships
  }

  def addRelationship(relationshipBuilder: RelationshipBuilder,
                      key: String,
                      value: String) = key match {
    case FlowModule.RelationshipName => relationshipBuilder.name(value)
    case _                                     => logger.warn("Relationship " + key + " not known")

  }

  def valueProperties(propertyDescriptorValueMap: JavaMap[PropertyDescriptor, String]): MutableMap[String, String] = {    
    propertyDescriptorValueMap.asScala.map(x => (x._1.getName, x._2))   
  }

  def successRelationship(relationships: JavaSet[Relationship]): Option[Relationship] = {
    return relationships.asScala.find(FlowModule.RelSuccessId == _.getName)
  }
}

class ProcessorUtils {

}